### PR TITLE
Feature/datachange capture transformer

### PR DESCRIPTION
**Pull request by lasseaj**
This PR introduces a transformer that captures data changes from a source table, based on a specified primary key. It is particularly relevant when working with a source that only provides full load data, and you only need to capture the changes in an upstream process.

### CODE REVIEW

File: tests/local/transformers/test_data_change_capture_transformer.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+from spetlrtools.testing import DataframeTestCase
+
+from spetlr.spark import Spark
+from spetlr.transformers.data_change_capture_transformer import (
+    DataChangeCaptureTransformer,
+)
+
+
+class CaptureHistoryTransformerTest(DataframeTestCase):
+    def setUp(self):
```
Comment (JeppeBlixen) [inline tests/local/transformers/test_data_change_capture_transformer.py:11]: I like the structure of your tests!

Comment (JeppeBlixen) [inline tests/local/transformers/test_data_change_capture_transformer.py:11]: You are missing a test of the logic relating the parameter cols_to_exclude

**Review by JeppeBlixen — COMMENTED** [review]
Nice work!

Missing some docs under spetlr/docs/transformations/

File: src/spetlr/transformers/data_change_capture_transformer.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def process_many(self, dataset: dataset_group) -> DataFrame:
+        df_source = dataset["source"]
+        df_target = dataset["target"]
+
+        self._warn_if_schema_differs(df_source, df_target)
+
+        # Add an md5 hash column to both DataFrames
+        df_source = self._md5_hash(df_source, "md5_source")
+        df_target = self._md5_hash(df_target, "md5_target")
+        df_joined = df_source.join(
```
[the remark this answers is not in the document]
Comment (lasseaj) [reply src/spetlr/transformers/data_change_capture_transformer.py:45]: I would say it's essential to matching hash-keys between source and target. This also helps in detecting new records where the source doesn't have a match."

File: src/spetlr/transformers/data_change_capture_transformer.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, primary_key, cols_to_exclude: List[str] = None):
+        self.primary_key = primary_key
+        if cols_to_exclude is None:
+            self.cols_to_exclude = []
+        elif primary_key in cols_to_exclude:
+            raise ValueError(f"The primary key '{primary_key}' should not be excluded.")
+        else:
+            self.cols_to_exclude = cols_to_exclude
+
+    def _md5_hash(self, df: DataFrame, colName: str):
```
[the remark this answers is not in the document]
Comment (lasseaj) [reply src/spetlr/transformers/data_change_capture_transformer.py:28]: Interesting - I'll reach out, then we can talk further about this

File: tests/local/utils/test_md5_hashcolumn.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import unittest
+
+from pyspark.sql.types import IntegerType, StringType, StructField, StructType
+
+from spetlr.spark import Spark
+from spetlr.utils.Md5HashColumn import Md5HashColumn
+
+
+class TestMd5HashColumn(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (lasseaj) [reply tests/local/utils/test_md5_hashcolumn.py:9]: I would say this part does it?
image

File: src/spetlr/utils/Md5HashColumn.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    param cols_to_exclude: Optional list of columns to exclude
+
+    return: DataFrame with md5 hash column
+    """
+
+    cols_to_exclude = cols_to_exclude or []
+
+    cols = [col for col in df.columns if col not in cols_to_exclude]
+    df = df.withColumn(colName, md5(concat_ws("||", *cols)))
+    return df
```
[the remark this answers is not in the document]
Comment (lasseaj) [reply src/spetlr/utils/Md5HashColumn.py:28]: Concat_ws handles NULLS

File: src/spetlr/transformers/data_change_capture_transformer.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        cols = [col for col in df.columns if col not in self.cols_to_exclude]
+        return df.withColumn(colName, md5(concat_ws("||", *cols)))
+
+    def _warn_if_schema_differs(self, df1: DataFrame, df2: DataFrame):
```
[the remark this answers is not in the document]
Comment (lasseaj) [reply src/spetlr/transformers/data_change_capture_transformer.py:33]: I have extended the warning message

File: tests/local/utils/test_md5_hashcolumn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (LauJohansson) [reply tests/local/utils/test_md5_hashcolumn.py:9]: You are right!
I meant what if a data row is (NONE, NONE, NONE, NONE) -> I just wanted to ensure that it also will generate a hash value as you expect

**Review by LauJohansson — APPROVED** [review]

**Review by LauJohansson — COMMENTED** [review]
I had only one more comment

File: tests/local/utils/test_md5_hashcolumn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lasseaj) [reply tests/local/utils/test_md5_hashcolumn.py:9]: I have added it as an extra row

**Review by LauJohansson — APPROVED** [review]

**Pull request merged**